packages/langium/src/grammar/workspace/document-builder.ts

```diff
@@ -0,0 +1,32 @@
+/******************************************************************************
+ * Copyright 2022 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import { LangiumSharedServices } from '../../services';
+import { interruptAndCheck } from '../../utils/promise-util';
+import { DefaultDocumentBuilder } from '../../workspace/document-builder';
+import { DocumentState } from '../../workspace/documents';
+import { Grammar } from '../generated/ast';
+import { LangiumGrammarServices } from '../langium-grammar-module';
+import { LangiumGrammarDocument } from './documents';
+
+export class LangiumGrammarDocumentBuilder extends DefaultDocumentBuilder {
+    constructor(services: LangiumSharedServices) {
+        super(services);
+        this.addTypeCollectionPhase();
+    }
+
+    private addTypeCollectionPhase() {
```

Comment:
I would not do this in a subclass because there's nothing to override in this case. Proposal:
 - Move this code to a function in `langium-grammar-module.ts`
 - Call the function in `createLangiumGrammarServices`